fusedrug/utils/cluster/cluster_using_mmseqs.py

```diff
@@ -0,0 +1,156 @@
+from fuse.data.ops import get_function_call_str
+import shutil
+import hashlib
+import subprocess
+from fuse.utils.file_io import save_text_file_safe, read_text_file
+import os
+
+def cluster(force_rebuild:bool = False, **kwargs):
+    """
+    A wrapped around cluster_impl() to allow caching 
+    see cluster_impl doc for details
+    
+    Args:
+        force_rebuild: rebuilds the data even if the hash indicates that it was already cached.
+    """
+
+    if 'output_dir' not in kwargs:
+        raise Exception('cluster() expects "output_dir" ')
+
+    output_dir = kwargs['output_dir']
+
+    os.makedirs(output_dir, exist_ok=True)
+
+    str_repr = get_function_call_str(
+        cluter_impl, 
+        _ignore_kwargs_names=['num_workers'], 
+        _include_code = False,
+        **kwargs
+        )
+
+    hash_value = hashlib.md5(str_repr.encode()).hexdigest()
+    already_created_hash_filename = os.path.join(output_dir, 'created_hash')
+    already_created_description_filename = os.path.join(output_dir, 'created_desc')
+
+    _rebuild_needed = False
+    if os.path.isfile(already_created_hash_filename):
+        existing_hash_value = read_text_file(already_created_hash_filename)
+        if hash_value == existing_hash_value:
+            if force_rebuild:
+                print(f'cluster::Hash value on disk matches the calculated one ({hash_value}), but force_rebuild was requested so rebuilding anyway.')
+                _rebuild_needed = True
+        else:
+            raise Exception('hash value on disk is a mismatch! please delete the directory manually and rerun.')
+            _rebuild_needed = True
+
+        existing_desc = read_text_file(already_created_description_filename)
+    else:
+        _rebuild_needed = True
+            
+    if _rebuild_needed and (not _rebuild_needed):
+        print(f'Hash value indicates that it was already built, not rebuilding. See description: {existing_desc}')
+        return
+
+    cluter_impl(**kwargs)
+
+    save_text_file_safe(already_created_description_filename, hash_value)
+    save_text_file_safe(already_created_hash_filename, str_repr)
+
+def cluter_impl(*, input_fasta_filename:str, output_dir:str, cluster_min_seqeunce_identity:float=0.4, cluster_method:str='cluster'):
+    """
+
+    Uses mmseqs to:
+
+    1. Remove 100% sequence identity duplicates
+    2. Cluster the remaining unique sequences into multiple clusters (e.g. with 70% sequence similary threshold )
+        This is useful for multiple purposes:
+        a. Creating cross validation/test splits that evaluate and demonstrate generalizability
+        b. Balanced sampling during training, sampling with inverse proportion to cluster size. (Similar to Class Balancing)        
+
+    Args:
+        input_fasta_filename: a fasta with an entry per molecular entity
+        output_dir: where the output will be generated
```

Comment:
Do we need to keep all the temp files created by mmseqs? Maybe we can delete it when done.
You should document here the name of the generated tsv file and its content.